<div style="border:solid Chocolate 2px; padding: 40px">
 
<b> Имран, привет!👋</b> 

Меня зовут Кирилл, я буду ревьюером твоего проекта. Я предлагаю общаться на «ты», но если привычнее на «вы», просто скажи об этом! 

Я буду оставлять комментарии в твоем проекте. Пожалуйста, не удаляй их, они могут быть полезны в случае повторной проверки.

Свои комментарии я буду обозначать <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:
  
<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным. Здесь же я могу давать советы и предложения.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div> 
 

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

     
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>

<br/>
<div class="alert alert-info">
Рекомендую для твоих комментариев использовать синий или любой другой цветной фон. Я постараюсь ничего не пропустить, но если комментарии будут выделены, то это мне очень поможет!
</div>
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой руководитель будет поступать также. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку.



<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Кирилл, привет! Приятно познакомиться, конечно можем на "ты".
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Cнова привет. Спасибо за заметное оформление твоих комментариев!
</div>


# Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: задача исследована
    - [x]  Исследован баланс классов
    - [x]  Изучены модели без учёта дисбаланса
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 3: учтён дисбаланс
    - [x]  Применено несколько способов борьбы с дисбалансом
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 4: проведено тестирование
- [x]  Удалось достичь *F1*-меры не менее 0.59
- [x]  Исследована метрика *AUC-ROC*

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Загрузка-и-знакомство-с-данными" data-toc-modified-id="Загрузка-и-знакомство-с-данными-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Загрузка и знакомство с данными</a></span></li><li><span><a href="#Проверим-на-дубли" data-toc-modified-id="Проверим-на-дубли-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Проверим на дубли</a></span></li><li><span><a href="#Приводим-наименования-столбцов-к-нижнему-регисту" data-toc-modified-id="Приводим-наименования-столбцов-к-нижнему-регисту-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Приводим наименования столбцов к нижнему регисту</a></span></li><li><span><a href="#Удаляем-лишние-столбцы" data-toc-modified-id="Удаляем-лишние-столбцы-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Удаляем лишние столбцы</a></span></li><li><span><a href="#Преобразуем-категориальные-параметры-и-изучаем-Geography" data-toc-modified-id="Преобразуем-категориальные-параметры-и-изучаем-Geography-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Преобразуем категориальные параметры и изучаем Geography</a></span></li><li><span><a href="#Работа-с-пропусками" data-toc-modified-id="Работа-с-пропусками-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Работа с пропусками</a></span></li><li><span><a href="#Масштабируем-признаки-и-формируем-итоговую-таблицу-для-работы" data-toc-modified-id="Масштабируем-признаки-и-формируем-итоговую-таблицу-для-работы-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span><del>Масштабируем признаки и формируем итоговую таблицу для работы</del></a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Посмотрим-на-распределение-классов-признака-exited" data-toc-modified-id="Посмотрим-на-распределение-классов-признака-exited-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Посмотрим на распределение классов признака exited</a></span></li><li><span><a href="#Попробуем-обучить-модель-на-данных-&quot;как-есть&quot;" data-toc-modified-id="Попробуем-обучить-модель-на-данных-&quot;как-есть&quot;-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Попробуем обучить модель на данных "как есть"</a></span><ul class="toc-item"><li><span><a href="#Разделим-выборки-на-тренировочную,-валидационную-и-тестовую." data-toc-modified-id="Разделим-выборки-на-тренировочную,-валидационную-и-тестовую.-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Разделим выборки на тренировочную, валидационную и тестовую.</a></span></li><li><span><a href="#Теперь-выполним-масштабирование-признаков" data-toc-modified-id="Теперь-выполним-масштабирование-признаков-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Теперь выполним масштабирование признаков</a></span></li><li><span><a href="#Проверим-несколько-моделей" data-toc-modified-id="Проверим-несколько-моделей-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>Проверим несколько моделей</a></span><ul class="toc-item"><li><span><a href="#Случайный-лес-с-подбором-гиперпараметров" data-toc-modified-id="Случайный-лес-с-подбором-гиперпараметров-2.2.3.1"><span class="toc-item-num">2.2.3.1&nbsp;&nbsp;</span>Случайный лес с подбором гиперпараметров</a></span></li><li><span><a href="#Дерево-решений-с-подбором-гиперпараметров" data-toc-modified-id="Дерево-решений-с-подбором-гиперпараметров-2.2.3.2"><span class="toc-item-num">2.2.3.2&nbsp;&nbsp;</span>Дерево решений с подбором гиперпараметров</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.2.3.3"><span class="toc-item-num">2.2.3.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Попробуем-рассчитать-модели-с-балансировкой-классов" data-toc-modified-id="Попробуем-рассчитать-модели-с-балансировкой-классов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Попробуем рассчитать модели с балансировкой классов</a></span><ul class="toc-item"><li><span><a href="#Неудачная-попытка-№1" data-toc-modified-id="Неудачная-попытка-№1-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Неудачная попытка №1</a></span></li><li><span><a href="#Переобучаем-Случайный-лес-с-учетом-баланса" data-toc-modified-id="Переобучаем-Случайный-лес-с-учетом-баланса-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Переобучаем Случайный лес с учетом баланса</a></span></li><li><span><a href="#Переобучаем-Дерево-решений-с-учетом-баланса" data-toc-modified-id="Переобучаем-Дерево-решений-с-учетом-баланса-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Переобучаем Дерево решений с учетом баланса</a></span></li><li><span><a href="#Переобучаем-Логистическую-регрессию-с-учетом-баланса" data-toc-modified-id="Переобучаем-Логистическую-регрессию-с-учетом-баланса-3.1.4"><span class="toc-item-num">3.1.4&nbsp;&nbsp;</span>Переобучаем Логистическую регрессию с учетом баланса</a></span></li><li><span><a href="#Сводим-итоги-в-одном-месте" data-toc-modified-id="Сводим-итоги-в-одном-месте-3.1.5"><span class="toc-item-num">3.1.5&nbsp;&nbsp;</span>Сводим итоги в одном месте</a></span></li></ul></li><li><span><a href="#Попробуем-увеличить-выборку-1" data-toc-modified-id="Попробуем-увеличить-выборку-1-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Попробуем увеличить выборку 1</a></span><ul class="toc-item"><li><span><a href="#Неудачная-попытка-№1" data-toc-modified-id="Неудачная-попытка-№1-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Неудачная попытка №1</a></span></li><li><span><a href="#Переобучаем-Случайный-лес-с-учетом-upsample" data-toc-modified-id="Переобучаем-Случайный-лес-с-учетом-upsample-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Переобучаем Случайный лес с учетом upsample</a></span></li><li><span><a href="#Переобучаем-Дерево-решений-с-учетом-upsample" data-toc-modified-id="Переобучаем-Дерево-решений-с-учетом-upsample-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Переобучаем Дерево решений с учетом upsample</a></span></li><li><span><a href="#Переобучаем-Логистическую-регрессию-с-учетом-upsample" data-toc-modified-id="Переобучаем-Логистическую-регрессию-с-учетом-upsample-3.2.4"><span class="toc-item-num">3.2.4&nbsp;&nbsp;</span>Переобучаем Логистическую регрессию с учетом upsample</a></span></li><li><span><a href="#Сводим-результаты-в-одном-месте" data-toc-modified-id="Сводим-результаты-в-одном-месте-3.2.5"><span class="toc-item-num">3.2.5&nbsp;&nbsp;</span>Сводим результаты в одном месте</a></span></li></ul></li><li><span><a href="#Пробуем-уменьшить-выборку-&quot;0&quot;" data-toc-modified-id="Пробуем-уменьшить-выборку-&quot;0&quot;-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Пробуем уменьшить выборку "0"</a></span><ul class="toc-item"><li><span><a href="#Неудачная-попытка-№1" data-toc-modified-id="Неудачная-попытка-№1-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Неудачная попытка №1</a></span></li><li><span><a href="#Переобучаем-Случайный-лес-с-учетом-downsample" data-toc-modified-id="Переобучаем-Случайный-лес-с-учетом-downsample-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span>Переобучаем Случайный лес с учетом downsample</a></span></li><li><span><a href="#Переобучаем-Дерево-решений-с-учетом-downsample" data-toc-modified-id="Переобучаем-Дерево-решений-с-учетом-downsample-3.3.3"><span class="toc-item-num">3.3.3&nbsp;&nbsp;</span>Переобучаем Дерево решений с учетом downsample</a></span></li><li><span><a href="#Переобучаем-Логистическую-регрессию-с-учетом-downsample" data-toc-modified-id="Переобучаем-Логистическую-регрессию-с-учетом-downsample-3.3.4"><span class="toc-item-num">3.3.4&nbsp;&nbsp;</span>Переобучаем Логистическую регрессию с учетом downsample</a></span></li><li><span><a href="#Сводим-итоги-в-одном-месте" data-toc-modified-id="Сводим-итоги-в-одном-месте-3.3.5"><span class="toc-item-num">3.3.5&nbsp;&nbsp;</span>Сводим итоги в одном месте</a></span></li></ul></li><li><span><a href="#Попробуем-также-поработать-с-порогами-логистической-регрессии" data-toc-modified-id="Попробуем-также-поработать-с-порогами-логистической-регрессии-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Попробуем также поработать с порогами логистической регрессии</a></span><ul class="toc-item"><li><span><a href="#Пробуем-на-исходной-модели-с-исходными-данными:" data-toc-modified-id="Пробуем-на-исходной-модели-с-исходными-данными:-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Пробуем на исходной модели с исходными данными:</a></span></li><li><span><a href="#Теперь-попробуем-на-сбалансированной-модели-и-с-исходными-данными:" data-toc-modified-id="Теперь-попробуем-на-сбалансированной-модели-и-с-исходными-данными:-3.4.2"><span class="toc-item-num">3.4.2&nbsp;&nbsp;</span>Теперь попробуем на сбалансированной модели и с исходными данными:</a></span></li><li><span><a href="#Осталась-исходная-модель-и-апсампл-данные:" data-toc-modified-id="Осталась-исходная-модель-и-апсампл-данные:-3.4.3"><span class="toc-item-num">3.4.3&nbsp;&nbsp;</span>Осталась исходная модель и апсампл данные:</a></span></li></ul></li><li><span><a href="#Итого,-взглянем-на-результаты-всех-попыток:" data-toc-modified-id="Итого,-взглянем-на-результаты-всех-попыток:-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Итого, взглянем на результаты всех попыток:</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍  Цели, задачи и план ясны.
</div>

# 0 Импорт библиотек и инструментов

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.dummy import DummyClassifier
from sklearn.utils import shuffle
from sklearn.metrics import (
    f1_score, mean_squared_error,
    accuracy_score, roc_curve,
    precision_score, recall_score,
    precision_recall_curve,
    roc_auc_score)

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
    
Если загружаем несколько функций из одного пакета, то удобнее записать это через запятую.
    
`from sklearn.metrics import f1_score, roc_auc_score`
    
А если через запятую будет больше трех позиций, в таком случае здесь было бы идеально привести оформление в соответствие с <a href="https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html">pep8</a> - не оставлять длинных строк. 

Такое оформление предпочтительнее:    
    
`from sklearn.metrics import (
    accuracy_score, 
    confusion_matrix, 
    f1_score,
    roc_auc_score
)` 
    
        
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Да, так выходит куда лаконичнее.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Ага, спасибо за доработку по этому моменту.
</div>


## Подготовка данных

### Загрузка и знакомство с данными

In [2]:
data = pd.read_csv('/datasets/Churn.csv')
display(data.info())
display(data.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


None

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

### Проверим на дубли

In [3]:
display(data.duplicated().sum())
display(data['CustomerId'].duplicated().sum())
display(data['RowNumber'].duplicated().sum())

0

0

0

**Данные загружены и первоначально изучены, можем накидать план по подготовке данных:**

1) Для начала привести наименования столбцов к нижнему регистру для удобства работы;\
2) Снести лишнее: столбцы RowNumber, CustomerId, Surname - для работы модели не нужны. Также детальнее посмотрим на Geography.\
3) Преобразовать категориальные признаки - Gender и, возможно, Geography.\
4) Tenure - кажется, что это значимый параметр для модели, но, именно здесь видим 9% пропусков. Необходимо взглянуть на столбец подробнее и понять, что с ним делать - дропнуть эти строки или как-то восполнить пропуски.


<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Супер. Ошибки определены верно.
</div>  


### Приводим наименования столбцов к нижнему регисту

In [4]:
data.rename(columns=str.lower, inplace=True)
data.head(1)

rownumber  customerid   surname  creditscore geography  gender  age  \
0          1    15634602  Hargrave          619    France  Female   42   

   tenure  balance  numofproducts  hascrcard  isactivemember  estimatedsalary  \
0     2.0      0.0              1          1               1        101348.88   

   exited  
0       1

Наименования столбцов приведены к нижнему регистру

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Колонки был в верблюжьем стиле. Мы убрали заглавные буквы, теперь, строго говоря, они без стиля. Можно было бы довести тогда до змеиного стиля с нижним подчеркиванием. Сейчас переделывать не нужно, потому что ниже придется менять названия переменных у столбцов.   
    
На будущее, для приведения к змеиному стилю, чтобы не делать вручную, можно использовать регулярные выражения и библиотеку re (import re). Например, так:


`df.columns = [re.sub(r'(?<!^)(?=[A-Z])', '_', i).lower() for i in df.columns]`
    
Подробнее про регулярные выражения можно почитать здесь: https://habr.com/ru/post/349860/    

Ну и еще один универсальный вариант − функция clean_column()  https://pypi.org/project/skimpy/ 

</div>


<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Лайк за полезный совет и информацию
</div>

### Удаляем лишние столбцы

In [5]:
data = data.drop(['rownumber', 'customerid', 'surname'], axis = 1)
data.head(1)

creditscore geography  gender  age  tenure  balance  numofproducts  \
0          619    France  Female   42     2.0      0.0              1   

   hascrcard  isactivemember  estimatedsalary  exited  
0          1               1        101348.88       1

Лишние столбцы удалены

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Всё верно

</div> 

### Преобразуем категориальные параметры и изучаем Geography

Сначала посмотрим на geography

In [6]:
display(data.groupby('geography')['exited'].count())
display(data.groupby('geography')['exited'].sum())

geography
France     5014
Germany    2509
Spain      2477
Name: exited, dtype: int64

geography
France     810
Germany    814
Spain      413
Name: exited, dtype: int64

Здесь можно увидеть, что в зависимости от страны отток различается, в частности, доля оттока по Германии намного выше.
Применим OHE для столбцов geography и gender. При этом, фиктивный признак у нас появится только для gender.

In [7]:
data_ohe = pd.get_dummies(data, drop_first = True)
data_ohe.rename(columns=str.lower, inplace=True)
data_ohe.head(1)

creditscore  age  tenure  balance  numofproducts  hascrcard  \
0          619   42     2.0      0.0              1          1   

   isactivemember  estimatedsalary  exited  geography_germany  \
0               1        101348.88       1                  0   

   geography_spain  gender_male  
0                0            0

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
    
А вот и нет. Фиктивный признак будет появляться в каждом случае. Смотри, если у нас три страны, убираем одну колонку со станой (например, Францию), остается Германия и Испания. Если видим нули у них, значит логически следует, что в колонке с Францией должна быть единица. То есть, одна из колонок оказывается избыточной.
    
И вот такая история, когда два признака в сумме дают корреляцию для третьего (когда можем вычислить из одних колонок, что точно лежит в других) может быть причиной мультиколлинеарности.
    
***
 
Здесь следует  избегать dummy-ловушки, удаляя одну из колонок и исключая мультиколлинеарность. ( `drop_first= True`)

Подробнее: https://www.codecamp.ru/blog/pandas-get-dummies/ 

</div>


<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Да, действительно, поправил.
</div>

Категориальные параметры преобразованы

### Работа с пропусками

In [8]:
display(data['tenure'].describe())
display(data.groupby('tenure')['tenure'].count())
display(data[data['tenure'] == 0].groupby('geography')['tenure'].count())
display(data[data['tenure'] == 0].groupby('geography')['exited'].sum())

count    9091.000000
mean        4.997690
std         2.894723
min         0.000000
25%         2.000000
50%         5.000000
75%         7.000000
max        10.000000
Name: tenure, dtype: float64

tenure
0.0     382
1.0     952
2.0     950
3.0     928
4.0     885
5.0     927
6.0     881
7.0     925
8.0     933
9.0     882
10.0    446
Name: tenure, dtype: int64

geography
France     187
Germany     98
Spain       97
Name: tenure, dtype: int64

geography
France     33
Germany    34
Spain      23
Name: exited, dtype: int64

Итак, в столбце tenure у нас 909 пропусков.\
Видим, что кол-во клиентов tenure от 1 до 9 составляет +/- 900, что близко к кол-ву пропусков.\
Кол-во клиентов с tenure 0 - всего 382, что выглядит достаточно странно, учитывая, что клиентов 1 - 952.\
Учитывая, что когорта клиентов, пришедших в определенное время может только уменьшаться (либо не меняться), кол-во клиентов 0 всего 382 вызывает подозрение, кроме этого - распределены они по странам равномерно. Врядли кол-во новых клиентов резко упало в разы, предполагаю, что пустые значения в данном случае - это 0, клиенты со сроком жизни до 1 года.

In [9]:
data_ohe['tenure'] = data_ohe['tenure'].fillna(5)
display(data_ohe.groupby('tenure')['tenure'].count())

tenure
0.0      382
1.0      952
2.0      950
3.0      928
4.0      885
5.0     1836
6.0      881
7.0      925
8.0      933
9.0      882
10.0     446
Name: tenure, dtype: int64

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Предположение, что пропуски там, где клиент пришел недавно достаточно неплохое. Но подтвердить его сейчас мы не можем. Поэтому лучше следовать принципу, что следующие данные будут такими же по распределению, как текущие, и лучшей тактикой минимизации ошибки будет значение из области наибольшей плотности распределения, не изменяющую структуры распределения признака. 
    
Доля пропусков не критично велика, возможно и решение с удалением. Однако данных у нас не много, я бы остановился на чем-то вроде медианы.

</div>


<div class="alert alert-info">
<b>Комментарий студента:</b>
    
В общем-то да, учитывая, что здесь +/- равномерное распределение, можем заменить на медиану.
</div>

Поменяли пустые значения на 0

### ~~Масштабируем признаки и формируем итоговую таблицу для работы~~

~~Масштабируем признаки на столбцах creditscore, estimatedsalary, age, tenure, balance~~

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Масштабирование необходимо сделать после разбиения на выборки: учимся на трейне, применяем масштабирование ко всем трем.


В реальном применении моделей у нас не будет возможности обучить скелер сразу и на тренировочной, и на рабочей выборке. Поэтому здесь воспроизводим реальную ситуацию - обучаемся только на тренировочной выборке, масштабируем обученным таким образом скелером все три выборки.
    

</div>


<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Спасибо, теперь картинка по этому процессу прояснилась - масштабирование переносим в раздел с разбивкой данных на выборки.
</div>

Итого, таблица для дальнейшей работы - df (для удобства)

In [10]:
df = data_ohe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   creditscore        10000 non-null  int64  
 1   age                10000 non-null  int64  
 2   tenure             10000 non-null  float64
 3   balance            10000 non-null  float64
 4   numofproducts      10000 non-null  int64  
 5   hascrcard          10000 non-null  int64  
 6   isactivemember     10000 non-null  int64  
 7   estimatedsalary    10000 non-null  float64
 8   exited             10000 non-null  int64  
 9   geography_germany  10000 non-null  uint8  
 10  geography_spain    10000 non-null  uint8  
 11  gender_male        10000 non-null  uint8  
dtypes: float64(3), int64(6), uint8(3)
memory usage: 732.5 KB


## Исследование задачи

### Посмотрим на распределение классов признака exited

In [11]:
display(df['exited'].value_counts(normalize = True))

0    0.7963
1    0.2037
Name: exited, dtype: float64

Видно, что есть явный дисбаланс, примерно 1 к 4.

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Ух, и здесь не совсем точно. Смотри, 0.2 к 0.8 делим оба числа на два =  0.1 к 0.4 умножаем на 10 =  1 к 4, а не 1 к 5. 

</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Соглашусь, нужно быть точнее в подаче информации. Поправил
</div>

### Попробуем обучить модель на данных "как есть"

#### Разделим выборки на тренировочную, валидационную и тестовую.

In [12]:
target = df['exited']
features = df.drop(['exited'] , axis=1)

features_train, features_valid1, target_train, target_valid1 = train_test_split(
    features, target, test_size=0.5, random_state=12345)
features_valid, features_test, target_valid, target_test = train_test_split(
    features_valid1, target_valid1, test_size=0.5, random_state=12345)

print(features_train.shape, target_train.shape, features_valid.shape, 
      target_valid.shape, features_test.shape, target_test.shape)

(5000, 11) (5000,) (2500, 11) (2500,) (2500, 11) (2500,)


Данные разделены на выборки. На данном этапе будем работать с train и valid.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 С разбиением на выборки согласен. Отлично, что выводим размерность.
</div>

#### Теперь выполним масштабирование признаков

In [13]:
numeric = ['creditscore', 'estimatedsalary', 'age', 'tenure', 'balance']
scaler = StandardScaler()

scaler.fit(features_train[numeric])

features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])
pd.options.mode.chained_assignment = None

display(features_train.head(2), features_valid.head(2), features_test.head(2))

/tmp/ipykernel_425/3702679547.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  features_train[numeric] = scaler.transform(features_train[numeric])
/opt/conda/lib/python3.9/site-packages/pandas/core/indexing.py:1738: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value[:, i].tolist(), pi)
/tmp/ipykernel_425/3702679547.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See

creditscore       age    tenure   balance  numofproducts  hascrcard  \
6331    -0.239180 -0.374637  1.091632  0.572035              1          1   
8201     0.685393  0.953729  1.823289  0.089503              1          0   

      isactivemember  estimatedsalary  geography_germany  geography_spain  \
6331               1        -0.140281                  1                0   
8201               1        -1.526589                  0                1   

      gender_male  
6331            1  
8201            0

creditscore       age    tenure   balance  numofproducts  hascrcard  \
7566    -0.810545 -0.279754 -0.371682  0.019486              1          1   
5696    -0.363841  0.194663  0.725803 -1.234200              2          0   

      isactivemember  estimatedsalary  geography_germany  geography_spain  \
7566               0         1.180220                  0                1   
5696               1        -1.468393                  0                0   

      gender_male  
7566            1  
5696            1

creditscore       age    tenure   balance  numofproducts  hascrcard  \
5403    -0.644330 -1.608120 -0.005853  0.780021              1          1   
4488     0.664616  0.479312  0.359975  1.259985              1          0   

      isactivemember  estimatedsalary  geography_germany  geography_spain  \
5403               1         1.079292                  0                1   
4488               0         0.603955                  0                0   

      gender_male  
5403            1  
4488            0

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Теперь все отлично.
</div>


#### Проверим несколько моделей

##### Случайный лес с подбором гиперпараметров

In [14]:
model_forest_best = None
best_result_f = 0
best_depth_f = 0
best_estimators_f = 0
best_roc_auc_f = 0
for depth in range(1, 50, 1):
    for estimators in range(1, 100, 10):
        model_f = RandomForestClassifier(n_estimators=estimators, max_depth=depth, random_state=12345)
        model_f.fit(features_train, target_train)
        predicted_valid_f = model_f.predict(features_valid)
        probabilities_f = model_f.predict_proba(features_valid)
        probabilities_one_f = probabilities_f[:, 1]
        if f1_score(target_valid, predicted_valid_f) > best_result_f:
            best_result_f = f1_score(target_valid, predicted_valid_f)
            best_roc_auc_f = roc_auc_score(target_valid, probabilities_one_f)
            best_depth_f = depth
            best_estimators_f = estimators
            model_forest_best = model_f
        
print('Лучший Случайный лес:', '| F1-мера:', best_result_f, '| ROC-AUC:', best_roc_auc_f, 
      '| Глубина:', best_depth_f, '| Кол-во деревьев', best_estimators_f)

Лучший Случайный лес: | F1-мера: 0.5682382133995036 | ROC-AUC: 0.8326109556335578 | Глубина: 13 | Кол-во деревьев 81


##### Дерево решений с подбором гиперпараметров

In [15]:
model_tree_best = None
best_result_t = 0
best_depth_t = 0
best_roc_auc_t = 0
for depth in range(1, 50, 1):
    model_t = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model_t.fit(features_train, target_train)
    predicted_valid_t = model_t.predict(features_valid)
    probabilities_t = model_t.predict_proba(features_valid)
    probabilities_one_t = probabilities_t[:, 1]
    if f1_score(target_valid, predicted_valid_t) > best_result_t:
        best_result_t = f1_score(target_valid, predicted_valid_t)
        best_roc_auc_t = roc_auc_score(target_valid, probabilities_one_t)
        best_depth_t = depth
        model_tree_best = model_t
        
print('Лучшее Дерево решений -', 'F1-мера:', best_result_t,
      '| ROC-AUC:', best_roc_auc_t, '| Глубина:', best_depth_t)

Лучшее Дерево решений - F1-мера: 0.5320665083135392 | ROC-AUC: 0.8007126932928564 | Глубина: 7


##### Логистическая регрессия


In [16]:
model_regr = LogisticRegression(random_state = 12345, solver = 'liblinear')
model_const = DummyClassifier(random_state = 12345)

model_regr.fit(features_train, target_train)
model_const.fit(features_train, target_train)

predict_regr_valid = model_regr.predict(features_valid)
predict_const_valid = model_const.predict(features_valid)

probabilities_r = model_regr.predict_proba(features_valid)
probabilities_one_r = probabilities_r[:, 1]

probabilities_c = model_const.predict_proba(features_valid)
probabilities_one_c = probabilities_c[:, 1]

print('Логистическая регрессия f1:', f1_score(target_valid, predict_regr_valid))
print('Конст. модель f1:', f1_score(target_valid, predict_const_valid))

print('Логистическая регрессия ROC-AUC:', roc_auc_score(target_valid, probabilities_one_r))
print('Конст. модель ROC-AUC:', roc_auc_score(target_valid, probabilities_one_c))

Логистическая регрессия f1: 0.31001371742112477
Конст. модель f1: 0.0
Логистическая регрессия ROC-AUC: 0.7297823182493753
Конст. модель ROC-AUC: 0.5


<div class="alert alert-warning">
<h2> Комментарий ревьюера 2<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь в дамми-модели я бы предсказал везде «1». Условно говоря, если предсказываем ноль, моделируем ситуацию, когда акцию с бонусами для удержания клиентов решили не проводить. Если предсказываем единицу − решили раздать бонусы вообще всем клиентам. Если в дамми-модели предсказываем по единице, можно подсчитать F1, она не будет зануляться.
    
`dummy_model = DummyClassifier(strategy='constant', constant=1)
dummy_model.fit(features_train, target_train)
f1_score(dummy_model.predict(features_test), target_test)`
    
Импорт: from sklearn.dummy import DummyClassifier

       
</div>


<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
     
roc_auc нужно считать по вероятностям классов. 
    
`probabilities = model.predict_proba(features)
probabilities_one = probabilities[:, 1]
print('Площадь ROC-кривой:', roc_auc_score(target, probabilities_one))`

Самое подробное объяснение метрики, что я видел на русском языке, можно посмотреть здесь: https://alexanderdyakonov.wordpress.com/2017/07/28/auc-roc-площадь-под-кривой-ошибок/ 

</div>



<div class="alert alert-block alert-danger">
<h2><S> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
В дереве и лесе важно подобрать гиперпараметры. В дереве - глубину, в лесе - минимум глубину и количество оценщиков (эстиматоров, деревьев). Смотреть на модели без подбора гиперпараметров не очень эффективно. Предлагаю убрать код для «деревянных» моделей (Леса, Дерева) выше.
    
Для Леса подойдет такой подход:
    
`for est in range(10, 101, 10):
    for depth in range(1, 14, 2):
        model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth)`
      


</div>


Лучший результат F1-меры у случайного леса, это говорит о том, что здесь наилучшее соотношение показателей полноты и качества предсказаний. Кроме этого, самый высокий accuracy и ROC-AUC также у случайного леса.

Попробуем улучшить модель Случайный лес:

 
<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь все верно. Единственное, нужно поправить вывод в принт результатов, сейчас там четыре числа без подписей. Важно подписать, что именно выводим. 

`print('F1-мера: ', best_result,  «...»` - по такой аналогии.
</div> 
  

Лучший результат при следующих гиперпараметрах: глубина - 21, кол-во деревьев - 81.

<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Несколько переосмыслил этот блок и немного изменил структуру
</div>

<div class="alert alert-info">
Итого соберем все результаты в одном месте:
</div>

In [17]:
print('Лучший Случайный лес:', '| F1-мера:', best_result_f, '| ROC-AUC:', best_roc_auc_f, 
      '| Глубина:', best_depth_f, '| Кол-во деревьев', best_estimators_f)
print('Лучшее Дерево решений -', 'F1-мера:', best_result_t,
      '| ROC-AUC:', best_roc_auc_t, '| Глубина:', best_depth_t)
print('Логистическая регрессия -', 'F1-мера:', f1_score(target_valid, predict_regr_valid),
      '| ROC-AUC:', roc_auc_score(target_valid, probabilities_one_r))
print('Константная модель -', 'F1-мера:', f1_score(target_valid, predict_const_valid),
      '| ROC-AUC:', roc_auc_score(target_valid, probabilities_one_c))

Лучший Случайный лес: | F1-мера: 0.5682382133995036 | ROC-AUC: 0.8326109556335578 | Глубина: 13 | Кол-во деревьев 81
Лучшее Дерево решений - F1-мера: 0.5320665083135392 | ROC-AUC: 0.8007126932928564 | Глубина: 7
Логистическая регрессия - F1-мера: 0.31001371742112477 | ROC-AUC: 0.7297823182493753
Константная модель - F1-мера: 0.0 | ROC-AUC: 0.5


**Выводы по разделу:**

1) Соотношение классов целевого признака - примерно 1 к 4 в пользу 0. Дисбаланс присутствует.

2) В текущих реалиях было протестировано 3 модели со следующими результатами:\
    - Логистическая регрессия: f1 = 0.31001371742112477, ROC-AUC = 0.7297823182493753;\
    - Дерево решений: f1 = 0.5320665083135392, ROC-AUC = 0.8007126932928564;\
    - Случайный лес: f1 = 0.5682382133995036, ROC-AUC = 0.8326109556335578.\
     
3) В итоге получили лучший f1 у Случайного леса = 0.5682382133995036, ROC-AUC = 0.8326109556335578 при n_estimators = 81 и max_depth = 13.

 
<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Важно подобрать гиперпараметры также в Дереве решений.
    
Здесь также стоит поправить по дисбалансу классов в комментарии, у нас 1 к 4. 
        
</div> 
  

<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Вывод поправил
</div>

## Борьба с дисбалансом

### Попробуем рассчитать модели с балансировкой классов

#### Неудачная попытка №1

model_regr_bal = LogisticRegression(solver = 'liblinear', class_weight = 'balanced', random_state = 12345)
model_tree_bal = DecisionTreeClassifier(random_state = 12345, class_weight = 'balanced')
model_forest_bal = RandomForestClassifier(random_state = 12345, class_weight = 'balanced')
model_forest_best_bal = RandomForestClassifier(n_estimators = 81, max_depth = 21, 
                                               random_state = 12345, class_weight = 'balanced')

model_regr_bal.fit(features_train, target_train)
model_tree_bal.fit(features_train, target_train)
model_forest_bal.fit(features_train, target_train)
model_forest_best_bal.fit(features_train, target_train)

predict_regr_valid_bal = model_regr_bal.predict(features_valid)
predict_tree_valid_bal = model_tree_bal.predict(features_valid)
predict_forest_valid_bal = model_forest_bal.predict(features_valid)
predict_forest_best_valid_bal = model_forest_best_bal.predict(features_valid)

print('Логистическая регрессия сбалансированная f1:', f1_score(target_valid, predict_regr_valid_bal))
print('Дерево решений сбалансированная f1:', f1_score(target_valid, predict_tree_valid_bal))
print('Случайный лес сбалансированная f1:', f1_score(target_valid, predict_forest_valid_bal))
print('Случайный лес улучшенная сбалансированная f1:', f1_score(target_valid, predict_forest_best_valid_bal))

print('Логистическая регрессия сбалансированная ROC-AUC:', roc_auc_score(target_valid, predict_regr_valid_bal))
print('Дерево решений сбалансированная ROC-AUC:', roc_auc_score(target_valid, predict_tree_valid_bal))
print('Случайный лес сбалансированная ROC-AUC:', roc_auc_score(target_valid, predict_forest_valid_bal))
print('Случайный лес улучшенная сбалансированная ROC-AUC:', roc_auc_score(target_valid, predict_forest_best_valid_bal))

Балансировка классов в данном случае помогла только логистической регрессии, f1 случайного леса и дерева решений были снижены.
Попробуем другой способ.

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Здесь изменились вводные  - теперь используем class_weight='balanced'. Необходимо заново обучать модели и заново находить лучшие гиперпараметры.

</div>


<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Здесь также делаю по новой, на всякий случай не удаляю первое решение. Ниже решение по новой структуре
</div>

#### Переобучаем Случайный лес с учетом баланса

In [18]:
model_forest_best_b = None
best_result_f_b = 0
best_depth_f_b = 0
best_estimators_f_b = 0
best_roc_auc_f_b = 0
for depth in range(1, 50, 1):
    for estimators in range(1, 100, 10):
        model_f_b = RandomForestClassifier(n_estimators=estimators, max_depth=depth, 
                                           random_state=12345, class_weight = 'balanced')
        model_f_b.fit(features_train, target_train)
        predicted_valid_f_b = model_f_b.predict(features_valid)
        probabilities_f_b = model_f_b.predict_proba(features_valid)
        probabilities_one_f_b = probabilities_f_b[:, 1]
        if f1_score(target_valid, predicted_valid_f_b) > best_result_f_b:
            best_result_f_b = f1_score(target_valid, predicted_valid_f_b)
            best_roc_auc_f_b = roc_auc_score(target_valid, probabilities_one_f_b)
            best_depth_f_b = depth
            best_estimators_f_b = estimators
            model_forest_best_b = model_f_b
        
print('Лучший Случайный лес balanced:', '| F1-мера:', best_result_f_b, '| ROC-AUC:', best_roc_auc_f_b, 
      '| Глубина:', best_depth_f_b, '| Кол-во деревьев', best_estimators_f_b)

Лучший Случайный лес balanced: | F1-мера: 0.591970121381886 | ROC-AUC: 0.8371458274041045 | Глубина: 9 | Кол-во деревьев 51


#### Переобучаем Дерево решений с учетом баланса

In [19]:
model_tree_best_b = None
best_result_t_b = 0
best_depth_t_b = 0
best_roc_auc_t_b = 0
for depth in range(1, 50, 1):
    model_t_b = DecisionTreeClassifier(max_depth=depth, random_state=12345, class_weight = 'balanced')
    model_t_b.fit(features_train, target_train)
    predicted_valid_t_b = model_t_b.predict(features_valid)
    probabilities_t_b = model_t_b.predict_proba(features_valid)
    probabilities_one_t_b = probabilities_t_b[:, 1]
    if f1_score(target_valid, predicted_valid_t_b) > best_result_t_b:
        best_result_t_b = f1_score(target_valid, predicted_valid_t_b)
        best_roc_auc_t_b = roc_auc_score(target_valid, probabilities_one_t_b)
        best_depth_t_b = depth
        model_tree_best_b = model_t_b
        
print('Лучшее Дерево решений balanced -', 'F1-мера:', best_result_t_b,
      '| ROC-AUC:', best_roc_auc_t_b, '| Глубина:', best_depth_t_b)

Лучшее Дерево решений balanced - F1-мера: 0.5620805369127517 | ROC-AUC: 0.8069522183221075 | Глубина: 6


#### Переобучаем Логистическую регрессию с учетом баланса

In [20]:
model_regr_b = LogisticRegression(random_state = 12345, solver = 'liblinear', class_weight = 'balanced')

model_regr_b.fit(features_train, target_train)

predict_regr_valid_b = model_regr_b.predict(features_valid)

probabilities_r_b = model_regr_b.predict_proba(features_valid)
probabilities_one_r_b = probabilities_r_b[:, 1]

print('Логистическая регрессия balanced -', 'F1-мера:', f1_score(target_valid, predict_regr_valid_b),
      '| ROC-AUC:', roc_auc_score(target_valid, probabilities_one_r_b))

Логистическая регрессия balanced - F1-мера: 0.4579310344827586 | ROC-AUC: 0.7338211112165196


#### Сводим итоги в одном месте

In [21]:
print('Лучший Случайный лес balanced:', '| F1-мера:', best_result_f_b, '| ROC-AUC:', best_roc_auc_f_b, 
      '| Глубина:', best_depth_f_b, '| Кол-во деревьев', best_estimators_f_b)
print('Лучшее Дерево решений balanced -', 'F1-мера:', best_result_t_b,
      '| ROC-AUC:', best_roc_auc_t_b, '| Глубина:', best_depth_t_b)
print('Логистическая регрессия balanced -', 'F1-мера:', f1_score(target_valid, predict_regr_valid_b),
      '| ROC-AUC:', roc_auc_score(target_valid, probabilities_one_r_b))

Лучший Случайный лес balanced: | F1-мера: 0.591970121381886 | ROC-AUC: 0.8371458274041045 | Глубина: 9 | Кол-во деревьев 51
Лучшее Дерево решений balanced - F1-мера: 0.5620805369127517 | ROC-AUC: 0.8069522183221075 | Глубина: 6
Логистическая регрессия balanced - F1-мера: 0.4579310344827586 | ROC-AUC: 0.7338211112165196


### Попробуем увеличить выборку 1

Здесь воспользуемся апсамплом, попробую выровнять выборки, увеличив "1" в 4 раза.

In [22]:
def upsample(features_train, target_train, repeat):
    features_zeros = features_train[target_train == 0]
    features_ones = features_train[target_train == 1]
    target_zeros = target_train[target_train == 0]
    target_ones = target_train[target_train == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

print(features_upsampled.shape, target_upsampled.shape)
print(target_upsampled.value_counts(normalize = True))

(8048, 11) (8048,)
1    0.50497
0    0.49503
Name: exited, dtype: float64


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 А здесь все верно по коэффициенту, действительно 4. 
    
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отлично, что используешь рукописную функцию, но здесь можно все оптимизировать, используя <a href="https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.SMOTE.html">SMOTE</a> для увеличения выборки и <a href="https://imbalanced-learn.org/stable/references/generated/imblearn.under_sampling.RandomUnderSampler.html">RandomUnderSampler</a>  для уменьшения. 
    
Апсемплинг делается очень просто:
    
`oversample = SMOTE(random_state=RANDOM_STATE)`

`features_train_up, target_train_up = oversample.fit_resample(features_train_enc, target_train)`
    
*предварительно импотрировав `from imblearn.over_sampling import SMOTE`

**Чтобы импортировать imblearn его сначала требуется проинсталлировать. через `!pip install imblearn -q`

В некоторых случаях качество может при применении SMOTE как улучшаться, так и ухудшаться в пределах статистической погрешности. Алгоритм не дублирует строки, а создает новые с помощью метода ближайших соседей. Можно попробовать посмотреть варианты с гиперпараметром `k_neighbors`. 

`oversample = SMOTE(random_state=RANDOM_STATE, k_neighbors=4)`

***
    
Если код выдаст ошибку, нужно перезагрузить ядро юпитера: Kernel - Restart на панели задач.
 
    
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Спасибо за интересный совет, обязательно попробую его в ближайшей задаче
</div>

#### Неудачная попытка №1

model_forest_best = RandomForestClassifier(n_estimators = 81, max_depth = 21, random_state = 12345)

model_regr.fit(features_upsampled, target_upsampled)
model_tree.fit(features_upsampled, target_upsampled)
model_forest.fit(features_upsampled, target_upsampled)
model_forest_best.fit(features_upsampled, target_upsampled)

predict_regr_valid_ups = model_regr.predict(features_valid)
predict_tree_valid_ups = model_tree.predict(features_valid)
predict_forest_valid_ups = model_forest.predict(features_valid)
predict_forest_best_valid_ups = model_forest_best.predict(features_valid)

print('Логистическая регрессия f1:', f1_score(target_valid, predict_regr_valid_ups))
print('Дерево решений f1:', f1_score(target_valid, predict_tree_valid_ups))
print('Случайный лес f1:', f1_score(target_valid, predict_forest_valid_ups))
print('Улучшенный случайный лес f1:', f1_score(target_valid, predict_forest_best_valid_ups))

print('Логистическая регрессия ROC-AUC:', roc_auc_score(target_valid, predict_regr_valid_ups))
print('Дерево решений ROC-AUC:', roc_auc_score(target_valid, predict_tree_valid_ups))
print('Случайный лес ROC-AUC:', roc_auc_score(target_valid, predict_forest_valid_ups))
print('Улучшенный случайный лес ROC-AUC:', roc_auc_score(target_valid, predict_forest_best_valid_ups))

По результатам апсамплинга f1 немного улучшилась, но все еще недостаточна.
Пробуем дальше.

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Заново подбираем гиперпараметры.
    
Еще раз напомню, ROC_AUC везде нужно будет не забыть поправить - по вероятностям классов расчитываем.
</div> 

<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Далее делаю по новой, код выше оставляю для истории
</div>

#### Переобучаем Случайный лес с учетом upsample

In [23]:
model_forest_best_up = None
best_result_f_up = 0
best_depth_f_up = 0
best_estimators_f_up = 0
best_roc_auc_f_up = 0
for depth in range(1, 50, 1):
    for estimators in range(1, 100, 10):
        model_f_up = RandomForestClassifier(n_estimators=estimators, max_depth=depth, 
                                           random_state=12345)
        model_f_up.fit(features_upsampled, target_upsampled)
        predicted_valid_f_up = model_f_up.predict(features_valid)
        probabilities_f_up = model_f_up.predict_proba(features_valid)
        probabilities_one_f_up = probabilities_f_up[:, 1]
        if f1_score(target_valid, predicted_valid_f_up) > best_result_f_up:
            best_result_f_up = f1_score(target_valid, predicted_valid_f_up)
            best_roc_auc_f_up = roc_auc_score(target_valid, probabilities_one_f_up)
            best_depth_f_up = depth
            best_estimators_f_up = estimators
            model_forest_best_up = model_f_up
        
print('Лучший Случайный лес upsampled:', '| F1-мера:', best_result_f_up, '| ROC-AUC:', best_roc_auc_f_up, 
      '| Глубина:', best_depth_f_up, '| Кол-во деревьев', best_estimators_f_up)

Лучший Случайный лес upsampled: | F1-мера: 0.5881081081081082 | ROC-AUC: 0.8197479880150526 | Глубина: 20 | Кол-во деревьев 61


#### Переобучаем Дерево решений с учетом upsample

In [24]:
model_tree_best_up = None
best_result_t_up = 0
best_depth_t_up = 0
best_roc_auc_t_up = 0
for depth in range(1, 50, 1):
    model_t_up = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model_t_up.fit(features_upsampled, target_upsampled)
    predicted_valid_t_up = model_t_up.predict(features_valid)
    probabilities_t_up = model_t_up.predict_proba(features_valid)
    probabilities_one_t_up = probabilities_t_up[:, 1]
    if f1_score(target_valid, predicted_valid_t_up) > best_result_t_up:
        best_result_t_up = f1_score(target_valid, predicted_valid_t_b)
        best_roc_auc_t_up = roc_auc_score(target_valid, probabilities_one_t_b)
        best_depth_t_up = depth
        model_tree_best_up = model_t_b
        
print('Лучшее Дерево решений upsampled -', 'F1-мера:', best_result_t_up,
      '| ROC-AUC:', best_roc_auc_t_up, '| Глубина:', best_depth_t_up)

Лучшее Дерево решений upsampled - F1-мера: 0.46043165467625896 | ROC-AUC: 0.659980473073396 | Глубина: 17


#### Переобучаем Логистическую регрессию с учетом upsample

In [25]:
model_regr_up = LogisticRegression(random_state = 12345, solver = 'liblinear', class_weight = 'balanced')

model_regr_up.fit(features_upsampled, target_upsampled)

predict_regr_valid_up = model_regr_b.predict(features_valid)

probabilities_r_up = model_regr_up.predict_proba(features_valid)
probabilities_one_r_up = probabilities_r_up[:, 1]

print('Логистическая регрессия upsampled -', 'F1-мера:', f1_score(target_valid, predict_regr_valid_up),
      '| ROC-AUC:', roc_auc_score(target_valid, probabilities_one_r_up))

Логистическая регрессия upsampled - F1-мера: 0.4579310344827586 | ROC-AUC: 0.7337825712298012


#### Сводим результаты в одном месте

In [26]:
print('Лучший Случайный лес upsampled:', '| F1-мера:', best_result_f_up, '| ROC-AUC:', best_roc_auc_f_up, 
      '| Глубина:', best_depth_f_up, '| Кол-во деревьев', best_estimators_f_up)
print('Лучшее Дерево решений upsampled -', 'F1-мера:', best_result_t_up,
      '| ROC-AUC:', best_roc_auc_t_up, '| Глубина:', best_depth_t_up)
print('Логистическая регрессия upsampled -', 'F1-мера:', f1_score(target_valid, predict_regr_valid_up),
      '| ROC-AUC:', roc_auc_score(target_valid, probabilities_one_r_up))

Лучший Случайный лес upsampled: | F1-мера: 0.5881081081081082 | ROC-AUC: 0.8197479880150526 | Глубина: 20 | Кол-во деревьев 61
Лучшее Дерево решений upsampled - F1-мера: 0.46043165467625896 | ROC-AUC: 0.659980473073396 | Глубина: 17
Логистическая регрессия upsampled - F1-мера: 0.4579310344827586 | ROC-AUC: 0.7337825712298012


### Пробуем уменьшить выборку "0"

Теперь воспользуемся даунсамплом - уменьшим выборку "0" в 4 раза, чтобы она была примерно равна выборке "1".

In [27]:
def downsample(features_train, target_train, fraction):
    features_zeros = features_train[target_train == 0]
    features_ones = features_train[target_train == 1]
    target_zeros = target_train[target_train == 0]
    target_ones = target_train[target_train == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    
    return features_downsampled, target_downsampled

features_downsampled, target_downsampled = downsample(features_train, target_train, 0.25)

#### Неудачная попытка №1

model_regr.fit(features_downsampled, target_downsampled)
model_tree.fit(features_downsampled, target_downsampled)
model_forest.fit(features_downsampled, target_downsampled)
model_forest_best.fit(features_downsampled, target_downsampled)

predict_regr_valid_ds = model_regr.predict(features_valid)
predict_tree_valid_ds = model_tree.predict(features_valid)
predict_forest_valid_ds = model_forest.predict(features_valid)
predict_forest_best_valid_ds = model_forest_best.predict(features_valid)

print('Логистическая регрессия f1:', f1_score(target_valid, predict_regr_valid_ds))
print('Дерево решений f1:', f1_score(target_valid, predict_tree_valid_ds))
print('Случайный лес f1:', f1_score(target_valid, predict_forest_valid_ds))
print('Улучшенный случайный лес f1:', f1_score(target_valid, predict_forest_best_valid_ds))

print('Логистическая регрессия ROC-AUC:', roc_auc_score(target_valid, predict_regr_valid_ds))
print('Дерево решений ROC-AUC:', roc_auc_score(target_valid, predict_tree_valid_ds))
print('Случайный лес ROC-AUC:', roc_auc_score(target_valid, predict_forest_valid_ds))
print('Улучшенный случайный лес ROC-AUC:', roc_auc_score(target_valid, predict_forest_best_valid_ds))

Даунсамплинг показал ухудшение результата.

Подводя итог, самым лучшим методом борьбы с дисбалансом классов в нашем случае оказался ансамплинг.\
Лучшая модель показала результат f1 = 0.578, который при даунсамплинге был 0.539, а при балансировке классов - 0.545. 

Без исправления дисбаланса, лучший результат f1 - 0.575

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
И здесь заново подбираем гиперпараметры: глубину в Дереве, глубину и количество деревьев в Лесе.
</div> 

<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Далее делаю по новой, код выше оставляю для истории
</div>

#### Переобучаем Случайный лес с учетом downsample

In [28]:
model_forest_best_ds = None
best_result_f_ds = 0
best_depth_f_ds = 0
best_estimators_f_ds = 0
best_roc_auc_f_ds = 0
for depth in range(1, 50, 1):
    for estimators in range(1, 100, 10):
        model_f_ds = RandomForestClassifier(n_estimators=estimators, max_depth=depth, 
                                           random_state=12345)
        model_f_ds.fit(features_downsampled, target_downsampled)
        predicted_valid_f_ds = model_f_ds.predict(features_valid)
        probabilities_f_ds = model_f_ds.predict_proba(features_valid)
        probabilities_one_f_ds = probabilities_f_ds[:, 1]
        if f1_score(target_valid, predicted_valid_f_up) > best_result_f_ds:
            best_result_f_ds = f1_score(target_valid, predicted_valid_f_ds)
            best_roc_auc_f_ds = roc_auc_score(target_valid, probabilities_one_f_ds)
            best_depth_f_ds = depth
            best_estimators_f_ds = estimators
            model_forest_best_ds = model_f_up
        
print('Лучший Случайный лес downsampled:', '| F1-мера:', best_result_f_ds, '| ROC-AUC:', best_roc_auc_f_ds, 
      '| Глубина:', best_depth_f_ds, '| Кол-во деревьев', best_estimators_f_ds)

Лучший Случайный лес downsampled: | F1-мера: 0.5363636363636364 | ROC-AUC: 0.81958888704424 | Глубина: 49 | Кол-во деревьев 91


#### Переобучаем Дерево решений с учетом downsample

In [29]:
model_tree_best_ds = None
best_result_t_ds = 0
best_depth_t_ds = 0
best_roc_auc_t_ds = 0
for depth in range(1, 50, 1):
    model_t_ds = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model_t_ds.fit(features_downsampled, target_downsampled)
    predicted_valid_t_ds = model_t_ds.predict(features_valid)
    probabilities_t_ds = model_t_ds.predict_proba(features_valid)
    probabilities_one_t_ds = probabilities_t_ds[:, 1]
    if f1_score(target_valid, predicted_valid_t_ds) > best_result_t_ds:
        best_result_t_ds = f1_score(target_valid, predicted_valid_t_ds)
        best_roc_auc_t_ds = roc_auc_score(target_valid, probabilities_one_t_ds)
        best_depth_t_ds = depth
        model_tree_best_ds = model_t_b
        
print('Лучшее Дерево решений downsampled -', 'F1-мера:', best_result_t_ds,
      '| ROC-AUC:', best_roc_auc_t_ds, '| Глубина:', best_depth_t_ds)

Лучшее Дерево решений downsampled - F1-мера: 0.5454545454545454 | ROC-AUC: 0.8123809213230877 | Глубина: 5


#### Переобучаем Логистическую регрессию с учетом downsample

In [30]:
model_regr_ds = LogisticRegression(random_state = 12345, solver = 'liblinear', class_weight = 'balanced')

model_regr_ds.fit(features_downsampled, target_downsampled)

predict_regr_valid_ds = model_regr_b.predict(features_valid)

probabilities_r_ds = model_regr_b.predict_proba(features_valid)
probabilities_one_r_ds = probabilities_r_ds[:, 1]

print('Логистическая регрессия downsampled -', 'F1-мера:', f1_score(target_valid, predict_regr_valid_ds),
      '| ROC-AUC:', roc_auc_score(target_valid, probabilities_one_r_ds))

Логистическая регрессия downsampled - F1-мера: 0.4579310344827586 | ROC-AUC: 0.7338211112165196


#### Сводим итоги в одном месте

In [31]:
print('Лучший Случайный лес downsampled:', '| F1-мера:', best_result_f_ds, '| ROC-AUC:', best_roc_auc_f_ds, 
      '| Глубина:', best_depth_f_ds, '| Кол-во деревьев', best_estimators_f_ds)
print('Лучшее Дерево решений downsampled -', 'F1-мера:', best_result_t_ds,
      '| ROC-AUC:', best_roc_auc_t_ds, '| Глубина:', best_depth_t_ds)
print('Логистическая регрессия downsampled -', 'F1-мера:', f1_score(target_valid, predict_regr_valid_ds),
      '| ROC-AUC:', roc_auc_score(target_valid, probabilities_one_r_ds))

Лучший Случайный лес downsampled: | F1-мера: 0.5363636363636364 | ROC-AUC: 0.81958888704424 | Глубина: 49 | Кол-во деревьев 91
Лучшее Дерево решений downsampled - F1-мера: 0.5454545454545454 | ROC-AUC: 0.8123809213230877 | Глубина: 5
Логистическая регрессия downsampled - F1-мера: 0.4579310344827586 | ROC-AUC: 0.7338211112165196


### Попробуем также поработать с порогами логистической регрессии

#### Пробуем на исходной модели с исходными данными:

In [32]:
model_regr.fit(features_train, target_train)

probabilities_valid = model_regr.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

for threshold in np.arange(0, 0.7, 0.05):
    predicted_valid = probabilities_one_valid > threshold
    f1score = f1_score(target_valid, predicted_valid)
    print("Порог = {:.2f} | F1 мера = {:.3f}".format(
        threshold, f1score))

Порог = 0.00 | F1 мера = 0.338
Порог = 0.05 | F1 мера = 0.365
Порог = 0.10 | F1 мера = 0.409
Порог = 0.15 | F1 мера = 0.441
Порог = 0.20 | F1 мера = 0.450
Порог = 0.25 | F1 мера = 0.454
Порог = 0.30 | F1 мера = 0.443
Порог = 0.35 | F1 мера = 0.418
Порог = 0.40 | F1 мера = 0.391
Порог = 0.45 | F1 мера = 0.356
Порог = 0.50 | F1 мера = 0.310
Порог = 0.55 | F1 мера = 0.268
Порог = 0.60 | F1 мера = 0.222
Порог = 0.65 | F1 мера = 0.155


<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

При обучении модель стремится достичь оптимального качества исходя из порога в 0.5. Подбор порога для повышения качества f1 метрики − здесь не самая эффективная мера, тем более на модели-аутсайдере. 
    
Но в целом иногда такое применяется при калибровке моделей.
        
</div> 
  
   


#### Теперь попробуем на сбалансированной модели и с исходными данными: 

In [33]:
probabilities_valid = model_regr_b.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

for threshold in np.arange(0, 0.7, 0.05):
    predicted_valid = probabilities_one_valid > threshold
    f1score = f1_score(target_valid, predicted_valid)
    print("Порог = {:.2f} | F1 мера = {:.3f}".format(
        threshold, f1score))

Порог = 0.00 | F1 мера = 0.338
Порог = 0.05 | F1 мера = 0.338
Порог = 0.10 | F1 мера = 0.346
Порог = 0.15 | F1 мера = 0.359
Порог = 0.20 | F1 мера = 0.373
Порог = 0.25 | F1 мера = 0.391
Порог = 0.30 | F1 мера = 0.409
Порог = 0.35 | F1 мера = 0.428
Порог = 0.40 | F1 мера = 0.444
Порог = 0.45 | F1 мера = 0.451
Порог = 0.50 | F1 мера = 0.458
Порог = 0.55 | F1 мера = 0.456
Порог = 0.60 | F1 мера = 0.458
Порог = 0.65 | F1 мера = 0.428


#### Осталась исходная модель и апсампл данные: 

In [34]:
model_regr.fit(features_upsampled, target_upsampled)


probabilities_valid = model_regr.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

for threshold in np.arange(0, 0.7, 0.05):
    predicted_valid = probabilities_one_valid > threshold
    f1score = f1_score(target_valid, predicted_valid)
    print("Порог = {:.2f} | F1 мера = {:.3f}".format(
        threshold, f1score))

Порог = 0.00 | F1 мера = 0.338
Порог = 0.05 | F1 мера = 0.338
Порог = 0.10 | F1 мера = 0.345
Порог = 0.15 | F1 мера = 0.360
Порог = 0.20 | F1 мера = 0.372
Порог = 0.25 | F1 мера = 0.391
Порог = 0.30 | F1 мера = 0.410
Порог = 0.35 | F1 мера = 0.428
Порог = 0.40 | F1 мера = 0.444
Порог = 0.45 | F1 мера = 0.452
Порог = 0.50 | F1 мера = 0.457
Порог = 0.55 | F1 мера = 0.457
Порог = 0.60 | F1 мера = 0.459
Порог = 0.65 | F1 мера = 0.436


Итого, логистическая регрессия все же показала худшие результаты в сравнении с другими моделями. И в нашем эксперменте, лучше всего она работает при балансировке классов (максимальный результат при минимальном пороге).

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь также не совсем понял про `максимальный результат при минимальном пороге`. Вижу, что хороший результат у нас как раз при пороге вокруг 0.5 
        
</div> 
  
   


<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Тут логика в том, что в случае балансировки классов модель показывает f1 более 0.45 при пороге 0.25, тогда как в других вариантах такой показатель f1 при порогах ближе к 0.5.
    Но, в целом все это бессмысленно, в данном случае я рассмотрел пороги Логистической регрессии чисто на всякий случай, не питая особых надежд.
</div>

<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
При тестировании мы уже смотрим на одну лучшую модель с лучшими гиперпараметрами. Тестирование - это применение чистых данных, которые наша модель еще не видела. Таким образом мы можем получить представление о том, какое качество модели мы можем ожидать в проде, в реальной жизни. 
    
Важно в тесте посмотреть на одну модель и с лучшими гиперпараметрами, подобранными ранее.
    
Важно до тестирования, в выводе по шагу исследования моделей определить эту самую лучшую модель и описать ее гиперпараметры, применяемый способ борьбы с дисбалансом классов (можно также добавить аргументацию, почему именно эту модель выбрали).

</div>


<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Чуть ниже выбираем лучшую модель
</div>

### Итого, взглянем на результаты всех попыток:

In [35]:
print('Balanced')
print('Лучший Случайный лес balanced:', '| F1-мера:', best_result_f_b, '| ROC-AUC:', best_roc_auc_f_b, 
      '| Глубина:', best_depth_f_b, '| Кол-во деревьев', best_estimators_f_b)
print('Лучшее Дерево решений balanced -', '| F1-мера:', best_result_t_b,
      '| ROC-AUC:', best_roc_auc_t_b, '| Глубина:', best_depth_t_b)
print('Логистическая регрессия balanced -', '| F1-мера:', f1_score(target_valid, predict_regr_valid_b),
      '| ROC-AUC:', roc_auc_score(target_valid, probabilities_one_r_b))

print('Upsampled')
print('Лучший Случайный лес upsampled:', '| F1-мера:', best_result_f_up, '| ROC-AUC:', best_roc_auc_f_up, 
      '| Глубина:', best_depth_f_up, '| Кол-во деревьев', best_estimators_f_up)
print('Лучшее Дерево решений upsampled -', '| F1-мера:', best_result_t_up,
      '| ROC-AUC:', best_roc_auc_t_up, '| Глубина:', best_depth_t_up)
print('Логистическая регрессия upsampled -', '| F1-мера:', f1_score(target_valid, predict_regr_valid_up),
      '| ROC-AUC:', roc_auc_score(target_valid, probabilities_one_r_up))

print('Downsampled')
print('Лучший Случайный лес downsampled:', '| F1-мера:', best_result_f_ds, '| ROC-AUC:', best_roc_auc_f_ds, 
      '| Глубина:', best_depth_f_ds, '| Кол-во деревьев', best_estimators_f_ds)
print('Лучшее Дерево решений downsampled -', '| F1-мера:', best_result_t_ds,
      '| ROC-AUC:', best_roc_auc_t_ds, '| Глубина:', best_depth_t_ds)
print('Логистическая регрессия downsampled -', '| F1-мера:', f1_score(target_valid, predict_regr_valid_ds),
      '| ROC-AUC:', roc_auc_score(target_valid, probabilities_one_r_ds))

Balanced
Лучший Случайный лес balanced: | F1-мера: 0.591970121381886 | ROC-AUC: 0.8371458274041045 | Глубина: 9 | Кол-во деревьев 51
Лучшее Дерево решений balanced - | F1-мера: 0.5620805369127517 | ROC-AUC: 0.8069522183221075 | Глубина: 6
Логистическая регрессия balanced - | F1-мера: 0.4579310344827586 | ROC-AUC: 0.7338211112165196
Upsampled
Лучший Случайный лес upsampled: | F1-мера: 0.5881081081081082 | ROC-AUC: 0.8197479880150526 | Глубина: 20 | Кол-во деревьев 61
Лучшее Дерево решений upsampled - | F1-мера: 0.46043165467625896 | ROC-AUC: 0.659980473073396 | Глубина: 17
Логистическая регрессия upsampled - | F1-мера: 0.4579310344827586 | ROC-AUC: 0.7337825712298012
Downsampled
Лучший Случайный лес downsampled: | F1-мера: 0.5363636363636364 | ROC-AUC: 0.81958888704424 | Глубина: 49 | Кол-во деревьев 91
Лучшее Дерево решений downsampled - | F1-мера: 0.5454545454545454 | ROC-AUC: 0.8123809213230877 | Глубина: 5
Логистическая регрессия downsampled - | F1-мера: 0.4579310344827586 | ROC-AUC

**По итогам подбора, лучшие показатели у модели Случайный лес с параметром class_weight = 'balanced'. Ее мы и будет тестировать далее**

## Тестирование модели

Попробуем прогнать все модели на тестовой выборке. Как показала практика выше, лучшие показатели у моделей, обученных на апсампл-выборках.

model_regr.fit(features_upsampled, target_upsampled)
model_tree.fit(features_upsampled, target_upsampled)
model_forest.fit(features_upsampled, target_upsampled)
model_forest_best.fit(features_upsampled, target_upsampled)

predict_regr_test_ups = model_regr.predict(features_test)
predict_tree_test_ups = model_tree.predict(features_test)
predict_forest_test_ups = model_forest.predict(features_test)
predict_forest_best_test_ups = model_forest_best.predict(features_test)

print('Логистическая регрессия f1:', f1_score(target_test, predict_regr_test_ups))
print('Дерево решений f1:', f1_score(target_test, predict_tree_test_ups))
print('Случайный лес f1:', f1_score(target_test, predict_forest_test_ups))
print('Улучшенный случайный лес f1:', f1_score(target_test, predict_forest_best_test_ups))

print('Логистическая регрессия ROC-AUC:', roc_auc_score(target_test, predict_regr_test_ups))
print('Дерево решений ROC-AUC:', roc_auc_score(target_test, predict_tree_test_ups))
print('Случайный лес ROC-AUC:', roc_auc_score(target_test, predict_forest_test_ups))
print('Улучшенный случайный лес ROC-AUC:', roc_auc_score(target_test, predict_forest_best_test_ups))

Получился достаточно интересный итог, получилось достичь f1 меры в 0,622 на модели Случайный лес. При этом, Улучшенный случайный лес показал результат немного хуже, как по f1 мере, так и по AUC-ROC.

AUC-ROC модели случайный лес получился 0.752 - это значительно выше 0.5 и может говорить о неплохой эффективности модели

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Соответственно, здесь правильнее будте исследовать лишь одну модель, которая выше показала себя как лучшая.
    

        
</div> 


<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Код выше оставляю для истории и приступаю к тестированию лучшей модели:
</div>

In [36]:
#model_forest_best_b
predicted_test = model_f_b.predict(features_test)
probabilities_test = model_f_b.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]
result_test = f1_score(target_test, predicted_test)
roc_auc_test = roc_auc_score(target_test, probabilities_one_test)
        
print('Лучший Случайный лес balanced:', '| F1-мера:', result_test, '| ROC-AUC:', roc_auc_test)

Лучший Случайный лес balanced: | F1-мера: 0.56590084643289 | ROC-AUC: 0.8577037292106293


Лучшая модель по итогам валидационных экспериментов оказалась недостаточно эффективной на тестовой выборке.\
Попробуем модель случайного леса upsampled:

In [37]:
#model_forest_best_up
predicted_test = model_forest_best_up.predict(features_test)
probabilities_test = model_forest_best_up.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]
result_test = f1_score(target_test, predicted_test)
roc_auc_test = roc_auc_score(target_test, probabilities_one_test)
        
print('Лучший Случайный лес upsampled:', '| F1-мера:', result_test, '| ROC-AUC:', roc_auc_test)

Лучший Случайный лес upsampled: | F1-мера: 0.6147368421052631 | ROC-AUC: 0.8527843261904131


Теперь целевой показатель достигнут. Лучшая модель - Случайный лес upsampled с "правильными" гиперпараметрами.

# Общий вывод

Итого, что было сделано и выявлено по ходу решения задачи:

1. Импортированы все необходимые библиотеки


2. Загружены и подготовлены данные:\
	1.1. Наименования столбцов приведены к нижнему регистру для удобства работы;\
	1.2. Удалены лишние столбцы: столбцы RowNumber, CustomerId, Surname - для работы модели не нужны.\
	1.3. Преобразованы категориальные признаки - Gender и Geography.\
	1.4. Обработаны пропуски в tenure (были заменены на медиану - 5).\
	1.5. Данные проверены на дубликаты.\
	1.6. Отмасштабированы количественные признаки.
    

3. Исследованы классы и обучены модели без исправления дисбаланса классов:\
	3.1. Соотношение классов - примерно 1 к 4 в пользу "0".\
	3.2. Датасет разделен на 3 выборки: тренировочную (50%), валидационную (25%) и тестовую (25%).\
	3.3. Обучены 3 модели: Логистическая регрессия (ЛР), Дерево решений (ДР) и Случайный лес (СЛ). Лучший показатель f1 - СЛ, 0.592. Лучший показатель ROC-AUC - СЛ, 0.837.\

    
4. Применено несколько способов работы с дисбалансом классов:\
	4.1. Были обучены модели с учетом баланса классов.\
	4.2. Были обучены модели на выборках с применением upsample.\
	4.3. Были обучены модели на выборках с применением downsample.\
    4.4. Лучший результат на валидационной выборке оказался у метода class_weight = balanced в целом, и у модели СЛ в частности.\
	4.5. Для ЛР отдельно были рассмотрены различные пороговые значения в разных вариациях самой модели (с балансом, без баланса, с апсамплом), однако все результаты значительно хуже других моделей.
    

5. Проведено тестирование моделей, обученных на апсампл-выборках: получилось достичь f1 меры в 0.615 на модели СЛ upsampled. При этом, лучшей моделью на валидационной выборке была СЛ balanced, но на тестовой выборке она показала результат значительно хуже, чем СЛ unsampled.


 <div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Выводы необходимо будет перепроверить и синхронизировать с решениями по коду.
       
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
    
Выводы поправил.
Большое спасибо за ценные советы и рекомендации!
</div>


<div class="alert alert-warning">
<h2> Комментарий ревьюера 2<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
    
Больше спасибо за внимательную доработку! 
    
Выше оставил комментарий по константной модели, там у тебя некорректный подсчет.
    
Здесь продублирую, чтобы не потерялось:

Здесь в дамми-модели я бы предсказал везде «1». Условно говоря, если предсказываем ноль, моделируем ситуацию, когда акцию с бонусами для удержания клиентов решили не проводить. Если предсказываем единицу − решили раздать бонусы вообще всем клиентам. Если в дамми-модели предсказываем по единице, можно подсчитать F1, она не будет зануляться.
    
`dummy_model = DummyClassifier(strategy='constant', constant=1)
dummy_model.fit(features_train, target_train)
f1_score(dummy_model.predict(features_test), target_test)`
    
Импорт: from sklearn.dummy import DummyClassifier

       
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проекту**
    
Спасибо за твой проект! Есть несколько важных моментов, которые необходимо поправить. Их ты найдешь в разноцветных комментариях.
    
**Отмечу положительные моменты проекта🚀**
    
* Хорошая структура проекта. Выбор моделей, порядок исследования.
* Есть пояснения к шагам, код с высокой читаемостью. 
* Логичность решений.
* Хороший результат на тестовой выборке.
    
**На что стоит обратить внимание🤔**
    
* Необходимо изменить подход к заполнению пропусков.
* Нужно масштабировать данные после разбиения на выборки
* Гиперпараметры моделям подбираем в каждом случае применения методики борьбы с дисбалансом и в исследовании моделей без учета дисбаланса.
* У всех моделей важно измерить ROC_AUC по вероятностям классов.
* Необходимо везде оставлять промежуточные выводы и дополнить итоговый вывод - перед тестированием обязательно описываем лучшую модель.
* В тестировании используем лишь одну модель - ту, которую нашли на шаге исследования моделей.
* Необходимо достичь качества модели F1>0.59 при честном тестировании одной модели (ну здесь ты уже нашел хороший вариант, просто нужно убрать лишнее)).
* Эти и остальные замечания я также написал по ходу проекта.
    
**Желаю удачи и жду твой проект на повторное ревью!**😉
    
    
</div>



<div style="border:solid Chocolate 2px; padding: 40px">

**Комментарий ко второй проверке**

Спасибо за доработку. Критичных комментариев нет. 
    
Если что-то пропустил или остались вопросы, напиши на телеграм @kriill

Желаю удачи в дальнейшей учебе!

</div>
